### PR TITLE
The great restructuring

### PR DESCRIPTION
**Pull request by PintarM**
This is a pull request to show some ideas for improving the internal structure of the package. It is not yet nowhere close to be merged yet, because in the end almost all internal functions will need to be (slightly) changed.

The main goal is to reduce the number of parameters passed around in internal functions and just use a few main ones (like field points and potential values). Then it should be easier to identify performance bottlenecks and improve them, because each function will be more independent of others. 

We can discuss this approach in the next meeting.

### CODE REVIEW

File: FindBounce/Kernel/FindBounce.m [hunk 1]
```diff
@@ -1661,19 +1661,18 @@
 	dim = OptionValue["Dimension"]/.Except[3|4]:>(Message[FindBounce::dim];Return[$Failed,Module]);
 	maxIteR = OptionValue["MaxRadiusIterations"]/.Except[_Integer?Positive]:>(Message[FindBounce::posint,"MaxRadiusIterations"];Return[$Failed,Module]);
 	maxItePath = OptionValue["MaxPathIterations"]/.Except[_Integer?NonNegative]:>(Message[FindBounce::nonnegint,"MaxPathIterations"];Return[$Failed,Module]);
-	(* TODO: Decision if extension method is used should be taken in one place only.
```
Comment (PintarM) [inline FindBounce/Kernel/FindBounce.m]: Sorry about this. I didn't understand the logic of how you decide if "extended" method is used or not.

File: FindBounce/Kernel/FindBounce.m [hunk 2]
```diff
@@ -1652,6 +1672,8 @@
 		potentialValues=potentialPoints;
 		extensionPB=False,
 		(* Otherwise we ask for value of potential at field points. *)
+		midPoint=findMidPoint[V,fields,{min1,min2},opts]/.($Failed:>Return[$Failed]);
```
Comment (PintarM) [inline FindBounce/Kernel/FindBounce.m:1702]: Nice! 👍

File: FindBounce/Kernel/FindBounce.m [hunk 3]
```diff
@@ -309,19 +310,14 @@
 Hessian calculation is part of another independent function, because the FindBounce options
 are independent. *)
 
-(* TODO: There is a problem with potentials given as functions with ?NumericQ argument
```
Comment (PintarM) [inline FindBounce/Kernel/FindBounce.m]: It is good that you deleted this now obsolete comment. 👍

File: FindBounce/Kernel/FindBounce.m [hunk 4]
```diff
@@ -309,19 +310,14 @@
...
-check. Function D or Grad can actually calculate numerical derivatives
-(probably some FDM in the background) but this can be quite slow. It would be better
-if "FiniteDifference" option value would be automatically chosen in such cases. *)
-
 FindBounce::gradmtd="Option value for Gradient should be Automatic, None, \"Symbolic\", \"FiniteDifference\" of custom function.";
 FindBounce::gradval=(
-	"The gradient of the potential is not well defined at some field point. "<> 
-	"Redefine the potential, choose option \"Gradient\"->\"FiniteDifference\".");
+	"The gradient of the potential is not well defined at some field points. "<> 
+	"Redefine the potential, choose option \"Gradient\"->\"FiniteDifference\" or \"MaxPathIterations\"->0.");
```
Comment (PintarM) [inline FindBounce/Kernel/FindBounce.m:316]: I like this attention to details.

File: FindBounce/Kernel/FindBounce.m [hunk 5]
```diff
@@ -1654,13 +1684,13 @@
 	];
 	
 	(* Checking of acceptable option values. If they are wrong function immediately returns $Failed.*)
-	bottomless = TrueQ@OptionValue["BottomlessPotential"];
-	pathTolerance = OptionValue["PathTolerance"]/.Except[_?NonNegative]:>(Message[FindBounce::posreal,"PathTolerance"];Return[$Failed,Module]);
-	actionTolerance = OptionValue["ActionTolerance"]/.Except[_?NonNegative]:>(Message[FindBounce::posreal,"ActionTolerance"];Return[$Failed,Module]);
-	dim = OptionValue["Dimension"]/.Except[3|4]:>(Message[FindBounce::dim];Return[$Failed,Module]);
-	maxIteR = OptionValue["MaxRadiusIterations"]/.Except[_Integer?Positive]:>(Message[FindBounce::posint,"MaxRadiusIterations"];Return[$Failed,Module]);
+	OptionValue["PathTolerance"]/.Except[_?NonNegative]:>(Message[FindBounce::posreal,"PathTolerance"];Return[$Failed,Module]);
```
Comment (PintarM) [inline FindBounce/Kernel/FindBounce.m:1687]: Good idea to reduce the number of local variables (for clarity).

File: FindBounce/Kernel/FindBounce.m [hunk 6]
```diff
@@ -1696,60 +1725,54 @@
...
-		
+	
 		(*Single Field Bounce.*)
-		{actionP,VL,v,a,b,pos,R,initialR} = If[
+		{actionP,potentialValues,bounce,initialR} = If[
 			Not[bottomless],
-			SingleFieldBounce[V,potentialPoints,Ns,noFields,\[Phi]L,dim,maxIteR,
-				actionTolerance,ansatzInitialR,initialR,rule,iter,
-				switchPath||iter==maxItePath
+			SingleFieldBounce[potentialValues,fieldPoints,
```
Comment (PintarM) [inline FindBounce/Kernel/FindBounce.m:1732]: Nice! This looks clearer and simpler to me.

**Review by PintarM — COMMENTED** [review]
Nice new documentation examples. 👍

File: FindBounce/Kernel/FindBounce.m (hunk 1 above)
Comment (vguada) [reply FindBounce/Kernel/FindBounce.m]: Hi Matevz, it is understandable., no problem 👍

File: FindBounce/Kernel/FindBounce.m (hunk 6 above)
Comment (vguada) [reply FindBounce/Kernel/FindBounce.m:1740]: I liked it as well. You had a good point. From now on I will try to have only functions with few arguments.

**Review by vguada — COMMENTED** [review]
Great! Good idea.

**Review by vguada — COMMENTED** [review]
Good idea! I like the examples, good job!

**Comment by vguada** [comment]
I think this branch can be merged to master already.

**Pull request merged**